flexmeasures/data/services/time_series.py

```diff
@@ -315,23 +315,63 @@ def set_bdf_source(bdf: tb.BeliefsDataFrame, source_name: str) -> tb.BeliefsData
 def drop_unchanged_beliefs(bdf: tb.BeliefsDataFrame) -> tb.BeliefsDataFrame:
     """Drop beliefs that are already stored in the database with an earlier belief time.
 
+    Also drop beliefs that are already in the data with an earlier belief time.
+
     Quite useful function to prevent cluttering up your database with beliefs that remain unchanged over time.
-    Only works on BeliefsDataFrames with a unique belief time and unique source.
     """
     if bdf.empty:
         return bdf
-    if len(bdf.lineage.belief_times) > 1:
-        raise NotImplementedError("Beliefs should share a unique belief time.")
-    if len(bdf.lineage.sources) > 1:
-        raise NotImplementedError("Beliefs should share a unique source.")
+
+    # Save the oldest ex-post beliefs explicitly, even if they do not deviate from the most recent ex-ante beliefs
+    ex_ante_bdf = bdf[bdf.belief_horizons > timedelta(0)]
+    ex_post_bdf = bdf[bdf.belief_horizons <= timedelta(0)]
+    if not ex_ante_bdf.empty and not ex_post_bdf.empty:
+        # Recursive function call
+        ex_ante_bdf = drop_unchanged_beliefs(ex_ante_bdf)
+        ex_post_bdf = drop_unchanged_beliefs(ex_post_bdf)
+        bdf = pd.concat([ex_ante_bdf, ex_post_bdf])
+        return bdf
+
+    # Remove unchanged beliefs from within the new data itself
+    index_names = bdf.index.names
+    bdf = (
+        bdf.sort_index()
+        .reset_index()
+        .drop_duplicates(
+            ["event_start", "source", "cumulative_probability", "event_value"],
+            keep="first",
+        )
+        .set_index(index_names)
+    )
+
+    return (
+        bdf.convert_index_from_belief_horizon_to_time()
+        .groupby(level=["belief_time", "source"], as_index=False)
+        .apply(_drop_unchanged_beliefs)
+    )
+
+
+def _drop_unchanged_beliefs(bdf: tb.BeliefsDataFrame) -> tb.BeliefsDataFrame:
```

Comment:
I'd rename this to `_drop_beliefs_unchanged_in_db` and also mention that it's usually better to not call it directly, but to call the public function first.